Major refactoring of SMC-NUTS code base

LJDevlin opened the pull request:
This PR is a major refactor of the SMC-NUTS codebase without change to its core functionality. Results obtained on this PR should match prior results.

The main contributions are:
1) Introduction of the samples class which contains variables and functions related to the samples themselves.
2) Scaling down of smc_sampler.py. This is still the main class to initiate the sampler but its main function is to run the main sampling loop, form estimates, output handling
3) Tidying up both run_experiments and plot_experiments by reducing the amount of repeated code
4) Moved the form estimates from tempering from its own class to the smc_sampler class.
5) Extra commenting
6) Reduced number of if-else statements

LJDevlin commented on smcnuts/samples/samples.py:
```diff
@@ -0,0 +1,224 @@
+        self.ess = 1 / np.sum(np.square(self.wn))
+
+
+    def resample_if_required(self):
```
Needs fixing:
indexes is not an argument
Consider returning samples and weights rather than setting here.

LJDevlin replied:
Fixed, and removed the returns. Instead we set the member variables in the function. I think the returns were there originally because we needed it that way when resampling post-processing, which has now been removed.

mjcarter95 commented on smcnuts/samples/samples.py:
```diff
@@ -0,0 +1,224 @@
+
+        return self.logw + p_logpdf_x_phi_new - p_logpdf_x_phi_old
+
+    def _non_assympototic_reweight(self):
```
should be `_non_asymptotic_reweight`

LJDevlin replied:
All typos fixed

mjcarter95 commented on smcnuts/samples/samples.py:
```diff
@@ -0,0 +1,224 @@
+        self.logw_new = self.reweight_strategy()
+
+            
+    def _assymptotic_reweight(self):
```
should be `_asymptotic`

LJDevlin replied:
typo fixed

mjcarter95 commented on smcnuts/samples/samples.py:
```diff
@@ -0,0 +1,224 @@
+        self.rng = rng
+
+        ## Set-up l-kernel if it is a function to be evaluated
+        if lkernel == "GaussianApproxLKernel":
```
"asymptotic" not "assymptotic"

LJDevlin replied:
typo fixed

mjcarter95 commented on smcnuts/samples/samples.py:
```diff
@@ -0,0 +1,224 @@
+
+        ## Set-up l-kernel if it is a function to be evaluated
+        if lkernel == "GaussianApproxLKernel":
+            self.lkernel = GaussianApproxLKernel(target=self.target, N=self.N)
```
Why did you decide against having the weight update as part of the L kernel class?

LJDevlin replied:
Two reasons:
1) It reduces repeated code since the weight update for the forwards proposal and Gaussian approximation are the the same for the weight update
2) It's your choice of proposal that defines the weight update i.e. accept-reject, not the L-kernel. It feels odd to me having the L-kernel be the class which defines how the weights are updated.

mjcarter95 replied:
Makes sense.

mjcarter95 commented on smcnuts/samples/samples.py:
```diff
@@ -0,0 +1,224 @@
+        
+        """
+        # Propogate particles through the forward kernel
+        self.r = self.forward_kernel.momentum_proposal.rvs(self.N)
```
Could be moved into `NUTSProposal`

LJDevlin replied:
I consciously decided to keep the momentum as a member of the samples class. The reason being momentum is a property of the samples themselves, and used in the weight update equation. 

I have moved gradient pre-calculation to the forwards proposal as it makes no sense to have it here.

mjcarter95 commented on smcnuts/samples/samples.py:
```diff
@@ -0,0 +1,224 @@
+        """
+        # Propogate particles through the forward kernel
+        self.r = self.forward_kernel.momentum_proposal.rvs(self.N)
+        grad_x = self.target.logpdfgrad(self.x, phi=self.phi_new)
```
Could be moved into `NUTSProposal`

LJDevlin replied:
See comment above

LJDevlin commented on smcnuts/smc_sampler.py:
```diff
@@ -168,131 +102,105 @@ def estimate(self, x, wn):
+        self.phi[self.K] = self.samples.phi_new
+        
+        # If using the asymptotic approach we calculate the estimates using the adjusted tempered weights
+        if(self.lkernel=="asymptoticLKernel"):
```
There is a hard assumption here that the asymptotic approach will always require tempering. I can make this more clear in the documentation but as tempering is more often used in the literature, this likely safe.

LJDevlin commented on smcnuts/tempering/adaptive_tempering.py:
Changed AdaptiveTempering to ESSTempering

LJDevlin commented on smcnuts/proposal/nuts_acc_rej.py:
fixed

LJDevlin commented:
Addressed all comments from last time.

I think this can go even further as discussed but this should be where we draw the line on this repository @mjcarter95. Can you check it produced repeatable results to the current main branch, also can you comment on anything relating to the structure of the code that is here. I think any additional refactoring of the structure should come later.

mjcarter95 commented:
@LJDevlin The results are equivalent to the current main branch (validated across all L kernels using two seeds).

I agree that the additional refactoring should come later. We could be more Pythonic in certain places (by adhering to PEP8 standards) and, we can remove reliance on strings (i.e. instead of `if lkernel == "GaussianApproxLKernel"` we can do `if instance(lkernel, GaussianApproximation)` which is neater IMO). These things can probably come later, the general structure looks good.
